statsd/client.py

```diff
@@ -138,14 +151,70 @@ def _send(self, data):
             # No time for love, Dr. Jones!
             pass
 
+    def pipeline(self):
+        return Pipeline(self)
+
+
+class TCPStatsClient(StatsClientBase):
+    """TCP version of StatsClient."""
+
+    def __init__(self, host='localhost', port=8125, prefix=None, timeout=None):
+        """Create a new client."""
+        self._host = host
+        self._port = port
+        self._timeout = timeout
+        self._prefix = prefix
+        self._tlocal = threading.local()
+        self._tlocal.sock = None
+
+    def _send(self, data):
+        """Send data to statsd."""
+        if not self._tlocal.sock:
+            self._connect()
+        try:
+            self._do_send(data)
+        except socket.error:
+            # try reconnecting and resending only once, so connections that
+            # have been up but died in the mean time will get re-established
+            self._reconnect_send(data)
+
+    def _reconnect_send(self, data):
+            self._close()
+            self._connect()
+            self._do_send(data)
+
+    def _connect(self):
+        family, _, _, _, addr = socket.getaddrinfo(
+            self._host, self._port, 0, socket.SOCK_STREAM)[0]
+        self._tlocal.sock = socket.socket(family, socket.SOCK_STREAM)
+        self._tlocal.sock.settimeout(self._timeout)
+        self._tlocal.sock.connect(addr)
+
+    def _do_send(self, data):
+        self._tlocal.sock.sendall(data.encode('ascii'))
```

Comment:
I think it's necessary to [terminate metrics with a newline](https://github.com/etsy/statsd/blob/master/docs/server.md) and unless I'm getting confused, it should be simple to add that here: `data.encode('ascii') + '\n'`.